ydb/core/blobstorage/vdisk/balance/balancing_actor.cpp

```diff
@@ -1,29 +1,191 @@
 #include "balancing_actor.h"
 #include "defs.h"
+#include "utils.h"
 
+#include <ydb/core/blobstorage/vdisk/common/vdisk_queues.h>
+#include <ydb/core/blobstorage/vdisk/repl/blobstorage_replbroker.h>
 
-namespace NKikimr {
 
+namespace NKikimr {
+namespace NBalancing {
     class TBalancingActor : public TActorBootstrapped<TBalancingActor> {
     private:
         std::shared_ptr<TBalancingCtx> Ctx;
-    public:
-        void Bootstrap() {
-            Become(&TThis::StateFunc);
+        TLogoBlobsSnapshot::TForwardIterator It;
+        TQueueActorMapPtr QueueActorMapPtr;
+        TActiveActors ActiveActors;
+        THashSet<TVDiskID> ConnectedVDisks;
+
+        TActorId SenderId;
+        TActorId DeleterId;
+
+        struct TStats {
+            bool SendCompleted = false;
+            bool DeleteCompleted = false;
+        };
+
+        TStats Stats;
+
+        void CreateVDisksQueues() {
+            QueueActorMapPtr = std::make_shared<TQueueActorMap>();
+            auto interconnectChannel = TInterconnectChannels::EInterconnectChannels::IC_BLOBSTORAGE_ASYNC_DATA;
+            const TBlobStorageGroupInfo::TTopology& topology = Ctx->GInfo->GetTopology();
+            NBackpressure::TQueueClientId queueClientId(
+                NBackpressure::EQueueClientType::Balancing, topology.GetOrderNumber(Ctx->VCtx->ShortSelfVDisk));
+
+            CreateQueuesForVDisks(*QueueActorMapPtr, SelfId(), Ctx->GInfo, Ctx->VCtx,
+                    Ctx->GInfo->GetVDisks(), Ctx->MonGroup.GetGroup(),
+                    queueClientId, NKikimrBlobStorage::EVDiskQueueId::GetAsyncRead,
+                    "DisksBalancing", interconnectChannel);
+        }
+
+        std::pair<TQueue<TPartInfo>, TQueue<TPartInfo>> CollectKeys(const TActorContext &ctx) {
+            TQueue<TPartInfo> sendOnMainParts, tryDeleteParts;
+
+            for (It.SeekToFirst(); It.Valid(); It.Next()) {
+                TPartsCollectorMerger merger(Ctx->GInfo->GetTopology().GType);
+                It.PutToMerger(&merger);
+
+                for (ui8 partIdx: PartsToSendOnMain(Ctx->GInfo->GetTopology(), Ctx->VCtx->ShortSelfVDisk, It.GetCurKey().LogoBlobID(), merger.Ingress)) {
+                    if (!merger.Parts[partIdx - 1].has_value()) {
+                        LOG_WARN_S(ctx, NKikimrServices::BS_VDISK_BALANCING,
```

Comment:
Ctx->VCtx->VDiskLogPrefix
or else it would be impossible to distinguish problem disks
applies to all log places